### ISSUE TITLE
[RFC]: A Triton operator dispatch mechanism through modified `CustomOp`

### ISSUE THREAD

**Issue by MengqingCao** [vllm-project/vllm#31467]
### Motivation.

Triton is becoming increasingly important in vLLM, and we've noticed its use in many models, quantization processes, and general workflows. Meanwhile, vLLM supports various backends. Typically, to achieve high performance, **different implementations of the Triton kernels** are used on different hardware, such as Ascend NPU. However, we've observed that vLLM currently lacks an effective operator dispatch mechanism for Triton to ensure that various backends can implement their own Triton kernels, which are then uniformly called by vLLM.

There are 3 ways of calling triton function now:

#### Through Attention Backend
Triton functions are called in `Attention` layer when the attention backend is specified as `TRITON_ATTN` or `TRITON_MLA`.

```python
current_platform.get_attn_backend_cls(...)
```

#### Through CustomOp
Some triton functions are included in other customops's forward pipeline, and they are put into `forward_cuda`, e.g., `causal_conv1d_fn` and `causal_conv1d_update` in `ShortConv`.

```python
class op1(CustomOp):
    def forward_cuda(kwargs):
        triton_fn(**kwargs)
```

#### Directly call
And there are others directly call triton functions in the normal pipeline.
  - some models derictly call triton functions in forward
    - Qwen3-Next
    - Kimi-Linear
    - ...
  - modelrunner v2
    - block table
    - input batch

Also, I notice that the implements are different form rocm and nvidia, algouth they are both cuda-alike platform.

```python
if current_platform.is_rocm():

    @triton.jit
    def round_int8(x):
        return tl.extra.hip.libdevice.round(x).to(tl.int8)

else:

    @triton.jit
    def round_int8(x):
        return tl.extra.cuda.libdevice.round(x).to(tl.int8)
```

### Proposed Change.

To solve the issues above, we propose to do the following changes:

<img width="1537" height="1346" alt="Image" src="https://github.com/user-attachments/assets/425dd6a4-b15c-4057-8aa8-25e4f563970b" />

1. Abstract a `CustomOpBase` class, which maintains funtions `register` and `register_oot`, which means all the instance of `CustomOpBase` could be registered in/out of vllm.
2. Seperate `CustomOp` and `CustomTritonOp`, we dispatch `CustomTritonOp` in a python func level, which pairs with the triton kernel. And the `CustomOp` keeps as is.
3. Refactor the exsiting triton kernels that are directly called without a python funtion warpping it, e.g., `eagle_prepare_inputs_padded_kernel`
4. Refactor the triton python functions to be hierit from `CustomTritonOp`, and optimize the current implement of triton kernel patching.

##### Code Change

```python
class CustomOpBase:
    """
    Base class for custom op. This class mainly offer the registry and dispatch function,
    and others must be overwrite in the sub classes.
    Dispatches the forward method to the appropriate backend.
    """

    op_registry: dict[str, Any] = {}
    op_registry_oot: dict[str, Any] = {}

    def __new__(cls, *args, **kwargs):
        try:
            op_name = cls.__name__
        except AttributeError:
            raise TypeError(
                f"Cannot instantiate '{cls.__name__}': its 'name' attribute "
                f"was not set, possibly because it was not decorated with "
                f"@CustomOp.register, or it's the CustomOp base class itself."
            ) from None

        if op_name not in cls.op_registry_oot:
            op_cls_to_instantiate = cls
        else:
            op_cls_to_instantiate = cls.op_registry_oot[op_name]
            logger.debug(
                "Instantiating custom op: %s using %s",
                op_name,
                str(op_cls_to_instantiate),
            )
        return super().__new__(op_cls_to_instantiate)

    def __init__(self, enforce_enable: bool = False):
        self._enforce_enable = enforce_enable
        self._forward_method = self.dispatch_forward()

    def forward(self, *args, **kwargs):
        return self._forward_method(*args, **kwargs)

    def forward_native(self, *args, **kwargs):
        raise NotImplementedError

    def forward_cuda(self, *args, **kwargs):
        raise NotImplementedError

    def forward_x(self, *args, **kwargs):
        raise NotImplementedError

    def forward_oot(self, *args, **kwargs):
        raise NotImplementedError

    def dispatch_forward(self):
        raise NotImplementedError

    # Decorator to register custom ops.
    @classmethod
    def register(cls, name: str):
        def decorator(op_cls):
            assert name not in cls.op_registry, f"Duplicate op name: {name}"
            op_cls.name = name
            cls.op_registry[name] = op_cls
            return op_cls

        return decorator

    @classmethod
    def register_oot(cls, _decorated_op_cls=None, name: str | None = None):
        def decorator(op_cls):
            reg_name = name if name is not None else cls.__name__
            assert reg_name not in cls.op_registry_oot, f"Duplicate op name: {reg_name}"
            op_cls.name = reg_name
            cls.op_registry_oot[reg_name] = op_cls
            return op_cls

        if _decorated_op_cls is None:
            # Called with parentheses: @CustomOP.register_oot()
            # or @CustomOP.register_oot(name="...")
            # So, _decorated_op_cls is None.
            # We return the actual decorator function.
            return decorator
        elif isinstance(_decorated_op_cls, type):  # Check if it's a class
            # Called without parentheses: @CustomOP.register_oot
            # The first argument is the class itself.
            # We call the 'decorator' function immediately with the class.
            return decorator(_decorated_op_cls)
        else:
            # Handle other unexpected cases if necessary
            raise TypeError("Decorator can only be applied to classes.")
```

```python
class CustomOp(nn.Module, CustomOpBase):
    """
    Base class for custom ops.
    Dispatches the forward method to the appropriate backend.
    """
    def dispatch_forward(self):
        ...

    @classmethod
    def enabled(cls) -> bool:
        ...

    @staticmethod
    def default_on() -> bool:
        ...
```

```python
class CustomTritonOp(CustomOpBase):
    """
    Base class for custom triton ops.
    Dispatches the forward method to the appropriate backend.
    """
    def dispatch_forward(self):
        ...

    @classmethod
    def enabled(cls) -> bool:
        ...

```

#### How to call a triton function?
Compared with the original implementation, We need to make definition  of class `TritonOp` and instance `triton_op` additionally.
```diff
@triton.jit
def triton_kernel():
    ...

def triton_python_fn():
    triton_kernel[grid](...)
    ...

+ class TritonOp(CustomTritonOp):
+     def forward():
+         triton_python_fn(...)

+ triton_op = TritonOp()
- output = triton_python_fn(...)
+ output = triton_op.forward(...)
```

When an oot platform want's to override an `CustomTritonOp`, e.g., `TritonOp`, we just goes same with the original `CustomOp`:
```python
@TritonOp.register_oot
class HPUUnquantizedFusedMoEMethod(TritonOp)
# or
CustomTritonOp.register_oot(name="TritonOp")
```

I have completed a rough code in vllm-ascend, and everthing goes well with this RFC, looking forward for more suggestions!

### Feedback Period.

A week

### CC List.

@WoosukKwon @youkaichao @jikunshang @DarkLight1337 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jikunshang** [comment]
Sounds reasonable to me. cc @tdoublep

**Comment by tjtanaa** [comment]
From my understanding, `CustomOp` is a place for developer to help user determine which is the best kernel to pick for the platform. The `CustomOp` class already has the logic to dispatch the ops based on platform, and each platform will decide their own logic which kernel to dispatch: cuda kernel OR triton kernel. 

So, for the case that you mentioned, where there are metadata difference between CUDA and ROCm, from what I could see is that each developer needs to invoke their specialized triton kernel in their respective dispatch function `forward_cuda/forward_hip`. At the `__init__` of the custom op class, those metadata information has to be determined already (to reduce overhead, as those metadata retrieval steps mostly are on CPU). Moreover triton kernels in many cases are not torch compile compatible especially if they are tunable, so they are not captured in graph, we will need to do `direct_register_customop`, the metadata retrievals overhead can add up.

Moreover, if we have `CustomOp` and `TritonCustomOp`, then we will need to name `class CausalConv1d(CustomTritonOp):` as `class TritonCausalConv1d(CustomTritonOp):`, since there can also be `class CausalConv1d(CustomOp):`

**Comment by MengqingCao** [comment]
> I would prefer to keep the `CustomOp` class as-is without introducing another `TritonCustomOp` abstraction. From my understanding, `CustomOp` is a place for developer to help user determine which is the best kernel to pick for the platform. The `CustomOp` class already has the logic to dispatch the ops based on platform, and each platform will decide their own logic which kernel to dispatch: cuda kernel OR triton kernel.
> 
> So, for the case that you mentioned, where there are metadata difference between CUDA and ROCm, from what I could see is that each developer needs to invoke their specialized triton kernel in their respective dispatch function `forward_cuda/forward_hip`. At the `__init__` of the custom op class, those metadata information has to be determined already (to reduce overhead, as those metadata retrieval steps mostly are on CPU). Moreover triton kernels in many cases are not torch compile compatible especially if they are tunable, so they are not captured in graph, we will need to do `direct_register_customop`, the metadata retrievals overhead can add up.
> 
> Moreover, if we have `CustomOp` and `TritonCustomOp`, then we will need to name `class CausalConv1d(CustomTritonOp):` as `class TritonCausalConv1d(CustomTritonOp):`, since there can also be `class CausalConv1d(CustomOp):`

just copying from https://github.com/vllm-project/vllm/pull/31521#issuecomment-3709633530, plz ignore this if you've already read it in the pr #31521

There are 2 main reasons for abstracting a new `TritonCustomOp`:

1. The enabling condition (Triton must be installed) for the triton kernel differs from that of the previous customop, resulting in different dispatch logic. Furthermore, the default implementation of the triton kernel is generally not a native torch implementation. Based on these differences, I think decoupling the two will make the operator semantics and code structure clearer.

2. In addition, we do not want `TritonCustomOp` to inherit from `torch.nn.Module`. Not inheriting from `torch.nn.Module` will reduce the number of guards and lower compilation complexity for `TritonCustomOp` instances.

Regarding your other comments:

1. The compatibility issue with torch compile: Triton kernels need to be manually wrapped under `torch.ops.vllm`, I think this issue is unrelated to whether it's in `CustomOp` or `TritonCustomOp`, but it's a very meaningful question. I'm interested in researching how tuable Triton kernels can effectively mask in a graph later.

2. Renaming `class CausalConv1d(CustomTritonOp)` to `class TritonCausalConv1d(CustomTritonOp)` makes big sense, and I will modify it according to this suggestion.

**Comment by MengqingCao** [comment]
> [@MengqingCao](https://github.com/MengqingCao) we will discuss here as this is the RFC thread.
> 
> Up till now, `forward_native` of `CustomOp` always guarantee torch compile compatible. So, it has always been torch, making it platform cross platform compatible by default. Triton is always installed on CUDA and ROCm as it is needed by torch compile. So, is this proposal suggested from the perspective of other than these two hardware (CUDA/ROCm)?
> 

Yep, such as Ascend NPU, and maybe intel guadi also meets the same issue.

> `CustomOp` semantics is quite clear that non-torch API functions are going to the custom `forward_xxx` path. So triton lives in the custom `forward_xxx` path for each hardware platform.
> 
> From code maintainability and `CustomOp` class understanding, I feel like we don't need to introduce another `TritonCustomOp`. Else we need to introduce another flag/switch to switch between `class CausalConv1d(CustomOp)` and `class TritonCausalConv1d(CustomTritonOp)` in `Layer` class. Different platform chooses to implement the same Op differently. There will be platform uses non-Triton op implementation. Adding another layer of dispatching logic to dispatch between `CustomOp` and `TritonCustomOp` in `layer.py` introduced is too much abstraction. So, if the design concern is from OOT hardware like, on Ascend NPU, you will call the triton kernel directly in `forward_oot` of vllm-ascend.
> 

Actually I thought of just implement a custom op directly in vllm-ascend, but I soon discovered there are a lot of triton kernels not be  warpped into a `CustomOp.forward`, plz refer to the 3 ways of calling triton functions described in the Motivition. Moreover, it seems the `CustomOp` remains more like a layer instead of an operator now, but triton kernel does not, I mean a triton kernel is not so heavy like the customops exsited. Therefore I think maybe it is better to modify the current customop slightly

> The point 2 is important for performance. Do you have any figures that we can use as reference how much overhead has the `nn.Module` incurred in general? We should highlight this and get more ideas on how to reduce this overhead in the whole repo if it is very large.

I just tried with a simple demo, as following:
```python

import operator
import torch
import torch.nn as nn
import torch._dynamo as dynamo

import triton
import triton.language as tl


class AtenModule(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
    def forward(self, x):
        out = torch.sin(x) + 1
        out = triton_add(out, out)
        return out

class AtenModuleTritonModule(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.add = TritonADD()
    def forward(self, x):
        out = torch.sin(x) + 1
        out = self.add(out)
        return out

@triton.jit
def add_kernel(
    x_ptr,
    y_ptr,
    out_ptr,
    n_elements,
    BLOCK: tl.constexpr,
):
    pid = tl.program_id(0)
    offs = pid * BLOCK + tl.arange(0, BLOCK)
    mask = offs < n_elements

    x = tl.load(x_ptr + offs, mask=mask)
    y = tl.load(y_ptr + offs, mask=mask)
    tl.store(out_ptr + offs, x + y, mask=mask)

def triton_add(
    x: torch.Tensor,
    y: torch.Tensor,
) -> torch.Tensor:
    out = torch.empty_like(x)
    n = x.numel()
    add_kernel[(16,)](x, y, out, n, BLOCK=1024)
    return out


class TritonADD(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
    def forward(self, input):
        out = triton_add(input, input)
        return out
        

def print_graph(fn, example_inputs, title):
    print("\n" + "=" * 80)
    print(title)
    print("=" * 80)

    gm, guards = dynamo.export(fn, *example_inputs)

    print("\n--- GraphModule code ---")
    print(gm.code)

    print("\n--- GraphModule graph ---")
    print(gm.graph)

    print("\n--- Guards ---")
    for g in guards:
        print(g)

def main():
    device = "cuda" if torch.cuda.is_available() else "cpu"
    x = torch.randn(1024, device=device)
    y = torch.randn(1024, device=device)

    A = AtenModule()
    A1 = AtenModuleTritonModule()
    print(50*"=")
    print(torch._dynamo.explain(A, x))
    print(50*"=")
    print_graph(A, (x,), "nn module + triton")
    print(50*"=")
    print(torch._dynamo.explain(A1, x))
    print(50*"=")
    print_graph(A1, (x,), "nn module + triton")


if __name__ == "__main__":
    main()

```
And the result is, without inheriting from `nn.Module`, there are 12 guards in total. When with inheriting from `nn.Module`, it increases to 27. I'll make a e2e test on Qwen3-Next later to measure the exact cost of this.

**Comment by tjtanaa** [comment]
> Thanks for discussion. I revisit the RFC and draft PR. here are my thoughts. I think we are making thing complex. **Any triton code should be a kernel instead of an operator.** What we need is a triton **kernel** dispatch mechanism for in tree and oot hardware/device(even different generation of hardware from same vendor may have differnet triton kernel implementation for best performance). My thought is adding a new class named `TritonKernelBase` to differentiate with existing `CustomOp`, they should be irrevelent logically, while they may have similar dispatch behavior.

Yes. I think the triton kernels should remain as an actual normal op. We shouldn't need to modify the `CustomOp` class. 
We can think of `CustomOp` is a framework level dispatching the best kernel for each platform, can be Non-Triton Kernel.

The `TritonOpBase` is to help Triton kernel developers to dispatch the best config for the triton kernel based on the platform. But this Triton op is OPTIONALLY called by the `CustomOp`. `CustomOp` could pick CUDA/HIP code that is faster than the Triton Kernel.

So, I think @jikunshang suggestion to view them as totally separate role.

**Comment by ProExpertProg** [comment]
Sorry for the late update, but this should be fully covered by vLLM IR: #32358. Additionally, the triton kernel registration can be similar to Helion kernel registration and management and we can reuse code if needed. Please comment on that doc/RFC if there is anything missing, but closing this RFC for now.
